rules/git/git_cached_repository.bzl

```diff
@@ -96,15 +96,48 @@ def _should_clone_repo(repo_ctx, repo_local_cache_path):
         ]
         st = repo_ctx.execute(args, quiet = True)
         if st.return_code == 0:
-            stripped_url = st.stdout.strip().replace("\n", "")
+            git_url = st.stdout.strip().replace("\n", "")
+            repo_rule_url = repo_ctx.attr.remote_url
 
             # In case remote url differ, a clone is required
-            return not stripped_url == repo_ctx.attr.remote_url
+            if not _is_repo_urls_equal(repo_ctx, git_url, repo_rule_url):
+                return True
         else:
-            fail(
-                "Failed to verify local git repository cached folder. name: {}, error: {}"
-                    .format(repo_ctx.name, st.stderr),
-            )
+            _log(repo_ctx, "identified an invalid remote origin url on local git index. name: {}, error: {}"
+                .format(repo_ctx.name, st.stderr))
+            return True
+
+        return not _is_valid_git_index(repo_ctx, repo_local_cache_path)
+
+def _is_repo_urls_equal(repo_ctx, git_url, rule_url):
+    same_repo_url = git_url == rule_url
+
+    if same_repo_url:
+        _log(repo_ctx, "git remote origin url and repository rule remote_url are the same. name: {}"
+            .format(repo_ctx.name))
+    else:
+        _log(repo_ctx, "git remote origin url and repository rule remote_url are different. name: {}"
+            .format(repo_ctx.name))
+
+    return same_repo_url
+
+def _is_valid_git_index(repo_ctx, repo_local_cache_path):
+    args = [
+        "git",
+        "-C",
+        repo_local_cache_path,
+        "log",
+        "-n",
+        "1",
+        "--pretty=format:%H",
```

Comment:
So this could have been any git command? you just expect it to fail if the repo is corrupted?

Reply:
Also I think this method is too long, I would break it so that each check is a different method, and here it only calls the checks.

Reply:
> So this could have been any git command? you just expect it to fail if the repo is corrupted?

This is the use case I'm trying to solve, devs sometimes encounter the following failure when the `git log..` command is executed:

```
fatal: your current branch 'master' does not have any commits yet
```

This usually happen when a repo git index was created but the Bazel build stops abruptly, it leaves an empty/invalid git index.

We rely on devs usage input, if a deeper index verification should take place due to other concerns we'll implement it, currently I believe this one is enough for letting devs to continue while we monitor this behavior.

> Also I think this method is too long, I would break it so that each check is a different method, and here it only calls the checks.

I agree, I'll break it apart 👍🏽

Reply:
Refactored the `should_clone` method, please have a look again 🙏🏽

Reply:
@or-shachar next time I'll tag you for review 😄

Reply:
We got a spy!

Reply:
Now the code is pretty 🌷